modular_skyrat/modules/delam_emergency_stop/code/scram.dm

```diff
@@ -0,0 +1,282 @@
+
+/obj/machinery/atmospherics/components/unary/delam_scram/Destroy()
+	QDEL_NULL(radio)
+	return ..()
+
+/obj/machinery/atmospherics/components/unary/delam_scram/proc/marry_sm()
+	my_sm = WEAKREF(GLOB.main_supermatter_engine)
+
+/obj/machinery/atmospherics/components/unary/delam_scram/update_icon_nopipes()
+	cut_overlays()
+	if(showpipe)
+		// everything is already shifted so don't shift the cap
+		add_overlay(get_pipe_image(icon, "inje_cap", initialize_directions, pipe_color))
+
+	if(!nodes[1] || !on || !is_operational)
+		icon_state = "inje_off"
+	else
+		icon_state = "inje_on"
+
+/obj/machinery/atmospherics/components/unary/delam_scram/process_atmos()
+	..()
+	if(!on || !is_operational)
+		return
+
+	var/turf/location = get_turf(loc)
+
+	if(isclosedturf(location))
+		return
+
+	var/datum/gas_mixture/air_contents = airs[1]
+
+	if(air_contents.temperature > 0)
+		var/transfer_moles = (air_contents.return_pressure() * volume_rate) / (air_contents.temperature * R_IDEAL_GAS_EQUATION)
+
+		if(!transfer_moles)
+			return
+
+		var/datum/gas_mixture/removed = air_contents.remove(transfer_moles)
+
+		location.assume_air(removed)
+
+		update_parents()
+
+/obj/machinery/atmospherics/components/unary/delam_scram/proc/panic_time(source, trigger_reason)
+	SIGNAL_HANDLER
+
+	if(on)
+		return
+
+	if(world.time - SSticker.round_start_time > 30 MINUTES)
+		playsound(src, 'sound/misc/compiler-failure.ogg', 100, FALSE, 15)
+		visible_message(span_danger("The [src] makes a series of sad beeps. The internal charge only lasts about 30 minutes... what a feat of engineering!"))
+		return
+
+	message_admins("Delam SCRAM activated!")
+	investigate_log("Delam SCRAM was activated by [trigger_reason ? "automatic safeties" : "manual intervention"]", INVESTIGATE_ATMOS)
+	radio.talk_into(src, "DELAMINATION SUPPRESSION SYSTEM FIRING IN 5 SECONDS. EVACUATE SUPERMATTER ENGINE ROOM!", emergency_channel)
+	SSpersistence.delam_highscore = SSpersistence.rounds_since_engine_exploded
+	SSpersistence.rounds_since_engine_exploded = BUTTON_PUSHED
+	for(var/obj/machinery/incident_display/sign as anything in GLOB.map_delamination_counters)
+		sign.update_delam_count(BUTTON_PUSHED)
+	addtimer(CALLBACK(src, PROC_REF(put_on_a_show)), 5 SECONDS)
+	// fight power with power
+	INVOKE_ASYNC(SSnightshift, TYPE_PROC_REF(/datum/controller/subsystem/nightshift, suck_light_power))
+
+/obj/machinery/atmospherics/components/unary/delam_scram/proc/put_on_a_show()
+	var/obj/machinery/power/supermatter_crystal/engine/angry_sm = my_sm?.resolve()
+	on = TRUE
+	alert_sound_to_playing('sound/misc/earth_rumble_distant3.ogg', override_volume = FALSE)
+	// good job at kneecapping the crystal, engineers
+	angry_sm.modify_filter(name = "ray", new_params = list(
+		color = SUPERMATTER_TESLA_COLOUR,
+	))
+	angry_sm.set_light_color(SUPERMATTER_TESLA_COLOUR)
+	// don't assume these dummies turned off the emitters
+	for(var/obj/machinery/power/emitter/shooty_boi in emitter_area)
```

Comment:
forgive me if im wrong, but doesnt iterating over an areas contents iterate over the entire world's contents in a really janky 

```
for (var/entry in world)
    if (entry.loc == entry)
      ...
       ...
```

i could be wrong but im also pretty sure im not so you really really really want to minimize area iterations

Reply:
I'm not sure if it does, but easily enough changed to signals in case.

Reply:
doing some searching in coderbus and yes it seems this is a certified byond moment

Reply:
lol rip